### PR TITLE
Migrate chain mask list config to full shard id list

### LINKED ISSUE #826: Fix chain mask list for slave-shard relationship
as of now our config is still using chain mask list to configure shard-slave mapping:

```
    """ Represent a mask of chains, basically matches all the bits from the right until the leftmost bit is hit.
    E.g.,
    mask = 1, matches *
    mask = 0b10, matches *0
    mask = 0b101, matches *01
    """
```

however it doesn't fit our current chain/shard ID decoding. we should migrate the chain mask list to be a list of full shard IDs indicating exactly what shards a slave should run.

### PR DESCRIPTION
**Pull request by lyhe18**
Add full_shard_id_list to slave to assign the shards it will run. In this case, each slave can run different number of shards.

resolves #826

### CODE REVIEW

File: quarkchain/cluster/cluster_config.py [hunk 1]
```diff
@@ -91,18 +91,19 @@ class SlaveConfig(BaseConfig):
     PORT = 38392
     WEBSOCKET_JSON_RPC_PORT = None
     ID = ""
-    CHAIN_MASK_LIST = None
+    # CHAIN_MASK_LIST = None
```
Comment (hanyunx) [inline quarkchain/cluster/cluster_config.py]: `chain_mask_list` part can be removed instead of commented if not needed

File: quarkchain/cluster/cluster_config.py [hunk 2]
```diff
@@ -91,18 +91,19 @@ class SlaveConfig(BaseConfig):
...
+    # def to_dict(self):
+    #     ret = super().to_dict()
+    #     ret["CHAIN_MASK_LIST"] = [m.value for m in self.CHAIN_MASK_LIST]
+    #     return ret
+    #
+    # @classmethod
+    # def from_dict(cls, d):
+    #     config = super().from_dict(d)
+    #     config.CHAIN_MASK_LIST = [ChainMask(v) for v in config.CHAIN_MASK_LIST]
+    #     return config
```
Comment (hanyunx) [inline quarkchain/cluster/cluster_config.py]: These two methods should also be modified for `full_shard_id_list`?

File: quarkchain/cluster/master.py [hunk 3]
```diff
@@ -433,7 +433,15 @@ class SlaveConnection(ClusterConnection):
...
-        self, env, reader, writer, master_server, slave_id, chain_mask_list, name=None
+        self,
+        env,
+        reader,
+        writer,
+        master_server,
+        slave_id,
+        full_shard_id_list,
+        name=None
+        # self, env, reader, writer, master_server, slave_id, chain_mask_list, name=None
```
Comment (hanyunx) [inline quarkchain/cluster/master.py]: can be removed

**Comment by ninjaahhh** [comment]
also, can you link this PR to the issue you want to solve?

File: quarkchain/core.py [hunk 4]
```diff
@@ -543,6 +543,20 @@ def has_overlap(self, chain_mask):
         return masks_have_overlap(self.value, chain_mask.value)
 
 
+class FullShardId(Serializable):
```
[the remark this answers is not in the document]
Comment (lyhe18) [reply quarkchain/core.py]: Sure, I will fix this~

[the remark this answers is not in the document]
Comment (ninjaahhh) [reply quarkchain/core.py]: should remove?

File: devnet/singularity/cluster_config.json [hunk 5]
```diff
@@ -380,35 +380,35 @@
       "HOST": "127.0.0.1",
       "PORT": 38000,
       "ID": "S0",
-      "CHAIN_MASK_LIST": [
-        4
+      "FULL_SHARD_ID_LIST": [
+        1, 262145
```
Comment (ninjaahhh) [inline devnet/singularity/cluster_config.json]: I would suggest using hex string like `"0x1"`, `"0x00010001"` then convert to integers when parsing the config

File: quarkchain/cluster/master.py [hunk 6]
```diff
@@ -876,23 +871,17 @@ def __has_all_shards(self):
         full_shard_ids = self.env.quark_chain_config.get_full_shard_ids()
         for slave, result in zip(slaves, results):
             # Verify the slave does have the same id and shard mask list as the config file
-            id, chain_mask_list = result
+            id, full_shard_id_list = result
+            # id, full_shard_id_list = result
```
Comment (ninjaahhh) [inline quarkchain/cluster/master.py]: remove

File: quarkchain/cluster/master.py [hunk 7]
```diff
@@ -876,23 +871,17 @@ def __has_all_shards(self):
...
             if id != slave.id:
                 Logger.error(
                     "Slave id does not match. expect {} got {}".format(slave.id, id)
                 )
                 self.shutdown()
-            if chain_mask_list != slave.chain_mask_list:
```
Comment (ninjaahhh) [inline quarkchain/cluster/master.py]: should have a similar check?

File: quarkchain/cluster/rpc.py [hunk 8]
```diff
@@ -41,48 +41,48 @@
 class Ping(Serializable):
     FIELDS = [
         ("id", PrependedSizeBytesSerializer(4)),
-        ("chain_mask_list", PrependedSizeListSerializer(4, ChainMask)),
+        ("full_shard_id_list", PrependedSizeListSerializer(4, FullShardId)),
```
Comment (ninjaahhh) [inline quarkchain/cluster/rpc.py]: should be integers?

**Comment by ninjaahhh** [comment]
we probably want to merge this to master branch instead of dev

File: devnet/singularity/cluster_config.json (hunk 5 above)
Comment (lyhe18) [reply devnet/singularity/cluster_config.json]: Done.

File: quarkchain/cluster/master.py (hunk 6 above)
Comment (lyhe18) [reply quarkchain/cluster/master.py]: Done.

File: quarkchain/cluster/rpc.py (hunk 8 above)
Comment (lyhe18) [reply quarkchain/cluster/rpc.py]: Done.

File: quarkchain/cluster/rpc.py [hunk 9]
```diff
@@ -41,48 +40,48 @@
 class Ping(Serializable):
     FIELDS = [
         ("id", PrependedSizeBytesSerializer(4)),
-        ("chain_mask_list", PrependedSizeListSerializer(4, ChainMask)),
+        ("full_shard_id_list", [uint32]),
```
Comment (ninjaahhh) [inline quarkchain/cluster/rpc.py]: this will break our serdes

File: quarkchain/cluster/slave.py [hunk 10]
```diff
@@ -707,46 +707,38 @@ def __init__(
...
         Logger.info("Closing connection with slave {}".format(self.id))
         return super().close_with_error(error)
 
     async def send_ping(self):
         # TODO: Send real root tip and allow shards to confirm each other
         req = Ping(
-            self.slave_server.id,
-            self.slave_server.chain_mask_list,
-            RootBlock(RootBlockHeader()),
+            self.slave_server.id, self.full_shard_id_list, RootBlock(RootBlockHeader()),
```
Comment (ninjaahhh) [inline quarkchain/cluster/slave.py]: `self.slave_server.full_shard_id_list`?

File: quarkchain/cluster/slave.py [hunk 11]
```diff
@@ -707,46 +707,38 @@ def __init__(
...
         self.ping_received_future.set_result(None)
 
-        return Pong(self.slave_server.id, self.slave_server.chain_mask_list)
+        return Pong(self.slave_server.id, self.slave_server.full_shard_id_list)
 
     # Blockchain RPC handlers
 
     async def handle_add_xshard_tx_list_request(self, req):
-        if req.branch not in self.shards:
+        if req.branch.value not in self.full_shard_id_list:
```
Comment (ninjaahhh) [inline quarkchain/cluster/slave.py]: is this change needed?

File: quarkchain/cluster/master.py [hunk 12]
```diff
@@ -876,23 +870,22 @@ def __has_all_shards(self):
...
+            if full_shard_id_list != slave.full_shard_id_list:
                 Logger.error(
-                    "Slave {} shard mask list does not match. expect {} got {}".format(
-                        slave.id, slave.chain_mask_list, chain_mask_list
+                    "Slave full shard id list does not match. expect {} got {}".format(
```
Comment (ninjaahhh) [inline quarkchain/cluster/master.py]: how about keeping slave ID?

**Comment by ninjaahhh** [comment]
this is a breaking change in terms of how we write the config (not of consensus). please give a review when you have time @qizhou

**Comment by ninjaahhh** [comment]
also cc @pingke if we want to make the config json universal regardless of client version

**Comment by ninjaahhh** [comment]
will update to make the config parsing backward-compatible, i.e. automatically converting chain mask list to full shard id list

File: quarkchain/cluster/cluster_config.py [hunk 13]
```diff
@@ -91,17 +91,21 @@ class SlaveConfig(BaseConfig):
...
+        # format to hex
+        ret["FULL_SHARD_ID_LIST"] = [
+            "0x{:08x}".format(i) for i in self.FULL_SHARD_ID_LIST
+        ]
         return ret
 
     @classmethod
     def from_dict(cls, d):
         config = super().from_dict(d)
-        config.CHAIN_MASK_LIST = [ChainMask(v) for v in config.CHAIN_MASK_LIST]
```
Comment (qizhou) [inline quarkchain/cluster/cluster_config.py]: We should maintain backward compability of CHAIN_MASK_LIST.  A simple way may just translate CHAIN_MASK_LIST + number of shards to FULL_SHARD_ID_LIST

Comment (ninjaahhh) [reply quarkchain/cluster/cluster_config.py]: done

File: quarkchain/cluster/cluster_config.py [hunk 14]
```diff
@@ -171,7 +203,7 @@ def __init__(self):
         slave_config = SlaveConfig()
         slave_config.PORT = 38000
         slave_config.ID = "S0"
-        slave_config.CHAIN_MASK_LIST = [ChainMask(1)]
+        slave_config.FULL_SHARD_ID_LIST = [2]
```
Comment (qizhou) [inline quarkchain/cluster/cluster_config.py]: Should FULL_SHARD_ID_LIST be [1] instead of [2] here?

File: quarkchain/cluster/cluster_config.py [hunk 15]
```diff
@@ -91,17 +91,49 @@ class SlaveConfig(BaseConfig):
...
+                "Can only have either FULL_SHARD_ID_LIST or CHAIN_MASK_LIST"
+            )
+        elif shard_ids:
+            # parse from hex to int
+            config.FULL_SHARD_ID_LIST = [int(h, 16) for h in config.FULL_SHARD_ID_LIST]
+        elif chain_mask:
+            # a simple way to be backward compatible with hard-coded shard ID
+            # e.g. chain mask 4 => 0x00000001, 0x00040001
+            # note that this only works if every chain has 1 shard only
+            check(all(chain.SHARD_SIZE == 1 for chain in cls._chains))
```
Comment (qizhou) [inline quarkchain/cluster/cluster_config.py]: could we have an optional argument with chains instead of a global class variable?  It seems that "chains" is known to the caller upon calling from_dict.  This also minimizes the scope of the "chains" variable.

File: quarkchain/cluster/cluster_config.py (hunk 14 above)
Comment (ninjaahhh) [reply quarkchain/cluster/cluster_config.py]: updated. though I don't see this default config being used any where

File: quarkchain/cluster/cluster_config.py (hunk 15 above)
Comment (ninjaahhh) [reply quarkchain/cluster/cluster_config.py]: done

File: quarkchain/cluster/cluster_config.py [hunk 16]
```diff
@@ -455,7 +497,10 @@ def from_dict(cls, d):
         config.QUARKCHAIN = QuarkChainConfig.from_dict(config.QUARKCHAIN)
         config.MONITORING = MonitoringConfig.from_dict(config.MONITORING)
         config.MASTER = MasterConfig.from_dict(config.MASTER)
-        config.SLAVE_LIST = [SlaveConfig.from_dict(s) for s in config.SLAVE_LIST]
+        config.SLAVE_LIST = [
+            SlaveConfig.from_dict(s, config.QUARKCHAIN.CHAINS)
```
Comment (qizhou) [inline quarkchain/cluster/cluster_config.py]: Oh, I think I may point to incorrect config variable (although CHAINS would work), which should be QUARKCHAIN.CHAINS_SIZE

Comment (ninjaahhh) [reply quarkchain/cluster/cluster_config.py]: since the backward-compatible conversion only works if shard size per chain is 1, I added an enforcing check as follows 

`check(all(chain.SHARD_SIZE == 1 for chain in chains))`

which needs the whole `QUARKCHAIN.CHAINS` config
